### PR TITLE
Avoid constant try/catch on a non-existing gateway for external cluster client

### LINKED ISSUE #8714: NullReferenceException in ClientClusterManifestProvider when all gateways are dead
**Orleans: 7.2.1**

When an external (read: not co hosted) cluster client has successfully started, and subsequently the remote cluster goes completely offline, a NRE is triggered every 5 seconds. 
This is what we see in the logging:

```
10:24:04.362 WARN  #045 Orleans.…e.ClientClusterManifestProvider Error trying to get cluster manifest from gateway null
System.NullReferenceException: Object reference not set to an instance of an object.
   at SystemTargetGrainId Orleans.Runtime.SystemTargetGrainId.Create(GrainType kind, SiloAddress address) in /_/src/Orleans.Core.Abstractions/IDs/SystemTargetGrainId.cs:line 44
   at async Task Orleans.Runtime.ClientClusterManifestProvider.RunAsync() in /_/src/Orleans.Core/Manifest/ClientClusterManifestProvider.cs:line 83
10:24:09.370 WARN  #024 Orleans.…e.ClientClusterManifestProvider Error trying to get cluster manifest from gateway null
System.NullReferenceException: Object reference not set to an instance of an object.
   at SystemTargetGrainId Orleans.Runtime.SystemTargetGrainId.Create(GrainType kind, SiloAddress address) in /_/src/Orleans.Core.Abstractions/IDs/SystemTargetGrainId.cs:line 44
   at async Task Orleans.Runtime.ClientClusterManifestProvider.RunAsync() in /_/src/Orleans.Core/Manifest/ClientClusterManifestProvider.cs:line 83
10:24:14.377 WARN  #040 Orleans.…e.ClientClusterManifestProvider Error trying to get cluster manifest from gateway null
System.NullReferenceException: Object reference not set to an instance of an object.
   at SystemTargetGrainId Orleans.Runtime.SystemTargetGrainId.Create(GrainType kind, SiloAddress address) in /_/src/Orleans.Core.Abstractions/IDs/SystemTargetGrainId.cs:line 44
   at async Task Orleans.Runtime.ClientClusterManifestProvider.RunAsync() in /_/src/Orleans.Core/Manifest/ClientClusterManifestProvider.cs:line 83
```

Note how it runs every 5 seconds.

Tracing the steps back, I find this in the `GatewayManager`:

```csharp
        /// <summary>
        /// Selects a gateway to use for a new bucket.
        ///
        /// Note that if a list provider delegate was given, the delegate is invoked every time this method is called.
        /// This method performs caching to avoid hammering the ultimate data source.
        ///
        /// This implementation does a simple round robin selection. It assumes that the gateway list from the provider
        /// is in the same order every time.
        /// </summary>
        /// <returns></returns>
        public SiloAddress GetLiveGateway()
        {
            List<SiloAddress> live = GetLiveGateways();
            int count = live.Count;
            if (count > 0)
            {
                lock (lockable)
                {
                    // Round-robin through the known gateways and take the next live one, starting from where we last left off
                    roundRobinCounter = (roundRobinCounter + 1) % count;
                    return live[roundRobinCounter];
                }
            }
            // If we drop through, then all of the known gateways are presumed dead
            return null;
        }
```
Note how this  returns null when all gateways are dead.

In `ClientClusterManifestProvider`, I see the following:

```csharp
        private async Task RunAsync()
        {
            try
            {
                var grainFactory = _services.GetRequiredService<IInternalGrainFactory>();
                var cancellationTask = _cancellation.Token.WhenCancelled();
                while (!_cancellation.IsCancellationRequested)
                {
                    var gateway = _gatewayManager.GetLiveGateway();
                    try
                    {
                        var provider = grainFactory.GetGrain<IClusterManifestSystemTarget>(SystemTargetGrainId.Create(Constants.ManifestProviderType, gateway).GrainId);
                        
                        // Rest of the code omitted for brevity
                        // ....
                    }
                    catch (Exception exception)
                    {
                        _logger.LogWarning(exception, "Error trying to get cluster manifest from gateway {Gateway}", gateway);
                        await Task.Delay(StandardExtensions.Min(_typeManagementOptions.TypeMapRefreshInterval, TimeSpan.FromSeconds(5)));
                    }
                }
            }
            finally
            {
                _initialized.TrySetResult(false);

                if (_logger.IsEnabled(LogLevel.Debug))
                {
                    _logger.LogDebug("Stopped refreshing cluster manifest");
                }
            }
        }
```

Note how the gateway (which can be null) is passed into `SystemTargetGrainId.Create`

This is the implementation of that method:

```csharp
        public static SystemTargetGrainId Create(GrainType kind, SiloAddress address) => new SystemTargetGrainId(new GrainId(kind, new IdSpan(address.ToUtf8String())));
```

At this point the NullReferenceException is thrown.

Context:

When we run our application stack in a single VM, or in a container cluster, sometimes service A that contains the client starts up faster than other services hosting the silos. In those cases, the membership table can be completely empty (in case we use a clean DB) or only contain dead silo records.

Also in scenarios where we deploy a new release, it's possible that all silos are offline while the client already starts up. I could communicate to our infra team that they must always start up the silos first, but that feels like not making enough effort to ensure this scenario just works. :-)

### PR DESCRIPTION
**Pull request by ledjon-behluli**
close #8714 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/8792)

### CODE REVIEW

File: src/Orleans.Core/Manifest/ClientClusterManifestProvider.cs [hunk 1]
```diff
@@ -94,6 +94,11 @@ private async Task RunAsync()
                     if (gateway is null || !_gatewayManager.IsGatewayAvailable(gateway))
                     {
                         gateway = _gatewayManager.GetLiveGateway();
+                        if (gateway is null)
+                        {
+                            continue;
```
Comment (ReubenBond) [inline src/Orleans.Core/Manifest/ClientClusterManifestProvider.cs:99]: Am I misreading or will this spin, calling `GetLiveGateway` in a tight loop? Perhaps this should contain a delay, similar to the other `continue` statements:

```suggestion
                            // Wait for a gateway to become available.
                            await Task.WhenAny(cancellationTask, Task.Delay(_typeManagementOptions.TypeMapRefreshInterval));
                            continue;
```

**Review by ReubenBond — CHANGES_REQUESTED** [review]
Just one change

File: src/Orleans.Core/Manifest/ClientClusterManifestProvider.cs (hunk 1 above)
Comment (ledjon-behluli) [reply src/Orleans.Core/Manifest/ClientClusterManifestProvider.cs:100]: absolutely, updated accordingly

**Review by ReubenBond — APPROVED** [review]

**Comment by ledjon-behluli** [comment]
@ReubenBond all good?

**Comment by ReubenBond** [comment]
Yep, all good - will merge when tests complete

**Pull request merged**